### DISCUSSION TITLE
How do you manage DTO / DAO layers with kysely?

### DISCUSSION THREAD

**Discussion by biryusal** [RobinBlomberg/kysely-codegen#216] [General]
Hi guys

I'm interested in how you manage DTO and DAO layers with `kysely-codegen`?

Since `codegen` returns interfaces, not classes, you can't just take and use them in the same `@nestjs/swagger` because it takes in a class to type methods.

Now I have this structure:
— **Controller**(args: `RequestDto`): `ResponseDto`
— **Service**(args: `RequestDto`): `ResponseDto`
— **Dao**(args: `PartOfDatabaseInterface`): `ResponseInterface`

And this is the most interesting part:
— **Dao**(args: `PartOfDatabaseInterface`): `ResponseInterface`

Since `codegen` generates interfaces, not classes, I need to rewrite everything that is in `ResponseInterface`, which automatically generates `kysely-codegen` — into exactly the same data, but wrapped in `class`

It looks like a hack, but I have not yet found another use to comply with the following practices:
— Have clearly delimited `Request DTO` that serialize and validate incoming data
— Have strict `Response DTO`, which always reflects with 100% accuracy what the backend returns
— Avoid duplicates of interfaces and classes for the cleanest code

@RobinBlomberg  — I sincerely apologize for mentioning you, but just so I can increase my chances that this post will not disappear, because it is the only one in discussions

**Comment by jakobsuckow** [comment]
Hi there, I am in the same boat. What was your solution in the end ?

**Comment by jakobsuckow** [comment]
I guess a custom ts-morph script might be the best approach here

**Comment by leourbina7r** [comment]
@jakobsuckow do you mind sharing what your solution with ts-morph looks like? Running into exactly the same issue.

**Comment by bakasmarius** [comment]
When I  started using ksysely (and codegen), I had  exactly the same problem, which I solved very easily with a simple command  - I run the generated  file through  a linter (ESLint or Oxlint)  with this rule: https://typescript-eslint.io/rules/consistent-type-definitions/  and interfaces are turned  into classes 😉
